[checklocks] Annotations ignored on generic structs

### Description

It appears `checklocks` (sha: [`69e0c7643d821f03ed797ea3911f6ab458e2a2d6`](https://github.com/google/gvisor/tree/69e0c7643d821f03ed797ea3911f6ab458e2a2d6/tools/checklocks)) fails to recognize annotations on generic types. For the following code:

```go
package example

import (
	"sync"
)

type X struct {
	mu sync.Mutex
	// +checklocks:mu
	m map[string]struct{}
}

func (x *X) Add(k string) {
	x.mu.Lock()
	defer x.mu.Unlock()
	x.m[k] = struct{}{}
}

type Y[K comparable] struct {
	mu sync.Mutex
	// +checklocks:mu
	m map[K]struct{}
}

func (y *Y[K]) Add(k K) {
	y.mu.Lock()
	defer y.mu.Unlock()
	y.m[k] = struct{}{}
}
```

`checklocks` returns the following:

```
go vet -vettool=.bin/checklocks ./example/...
# example.com/example
example/x.go:22:2: may require checklocks annotation for mu, used with lock held 100% of the time
```

### Steps to reproduce

_No response_

### runsc version

_No response_

### docker version (if using docker)

_No response_

### uname

_No response_

### kubectl (if using Kubernetes)

_No response_

### repo state (if built from source)

_No response_

### runsc debug logs (if available)

_No response_

--- a/tools/checklocks/BUILD
+++ b/tools/checklocks/BUILD
@@ -12,6 +12,7 @@ go_library(
         "annotations.go",
         "checklocks.go",
         "facts.go",
+        "importfacts.go",
         "state.go",
     ],
     nogo = False,

--- a/tools/checklocks/analysis.go
+++ b/tools/checklocks/analysis.go
@@ -221,7 +221,7 @@ func (pc *passContext) checkGuards(inst almostInst, from ssa.Value, accessObj ty
 	)
 
 	// Load the facts for the object accessed.
-	pc.pass.ImportObjectFact(accessObj, &lgf)
+	pc.importLockGuardFacts(accessObj, &lgf)
 
 	// Check guards held.
 	for guardName, fgr := range lgf.GuardedBy {
@@ -356,7 +356,7 @@ func (pc *passContext) checkCall(call callCommon, lff *lockFunctionFacts, ls *lo
 	// "invoke" mode: Method is non-nil, and Value is the underlying value.
 	if fn := call.Common().Method; fn != nil {
 		var nlff lockFunctionFacts
-		pc.pass.ImportObjectFact(fn, &nlff)
+		pc.importLockFunctionFacts(fn, &nlff)
 		nlff.Ignore = nlff.Ignore || lff.Ignore // Inherit ignore.
 		pc.checkFunctionCall(call, fn, &nlff, ls)
 		return
@@ -385,9 +385,10 @@ func (pc *passContext) checkCall(call callCommon, lff *lockFunctionFacts, ls *lo
 			Ignore: lff.Ignore, // Inherit ignore.
 		}
 		if obj := fn.Object(); obj != nil {
-			pc.pass.ImportObjectFact(obj, &nlff)
+			tf := obj.(*types.Func)
+			pc.importLockFunctionFacts(tf, &nlff)
 			nlff.Ignore = nlff.Ignore || lff.Ignore // See above.
-			pc.checkFunctionCall(call, obj.(*types.Func), &nlff, ls)
+			pc.checkFunctionCall(call, tf, &nlff, ls)
 		} else {
 			// Anonymous functions have no facts, and cannot be
 			// annotated.  We don't check for violations using the
@@ -869,7 +870,7 @@ func (pc *passContext) checkFunction(call callCommon, fn *ssa.Function, lff *loc
 func (pc *passContext) checkInferred() {
 	for obj, oo := range pc.observations {
 		var lgf lockGuardFacts
-		pc.pass.ImportObjectFact(obj, &lgf)
+		pc.importLockGuardFacts(obj, &lgf)
 		for other, count := range oo.counts {
 			// Is this already a guard?
 			if _, ok := lgf.GuardedBy[other.Name()]; ok {

--- a/tools/checklocks/checklocks.go
+++ b/tools/checklocks/checklocks.go
@@ -79,6 +79,7 @@ type passContext struct {
 
 // observationsFor retrieves observations for the given object.
 func (pc *passContext) observationsFor(obj types.Object) *objectObservations {
+	obj = originObject(obj)
 	if pc.observations == nil {
 		pc.observations = make(map[types.Object]*objectObservations)
 	}
@@ -187,11 +188,12 @@ func run(pass *analysis.Pass) (any, error) {
 		// by named functions in the package, and they are analyzing
 		// inline on every call. Thus we skip the analysis here. They
 		// will be hit on calls, or picked up in the pass below.
-		if obj := fn.Object(); obj == nil {
+		obj := fn.Object()
+		if obj == nil {
 			continue
 		}
 		var lff lockFunctionFacts
-		pc.pass.ImportObjectFact(fn.Object(), &lff)
+		pc.importLockFunctionFacts(obj.(*types.Func), &lff)
 
 		// Check the basic blocks in the function.
 		pc.checkFunction(nil, fn, &lff, nil, false /* force */)

--- a/tools/checklocks/importfacts.go
+++ b/tools/checklocks/importfacts.go
@@ -1,0 +1,59 @@
+// Copyright 2026 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checklocks
+
+import "go/types"
+
+// originObject returns obj's origin (for instantiated generic objects) when
+// available, otherwise it returns obj unchanged.
+func originObject(obj types.Object) types.Object {
+	switch obj := obj.(type) {
+	case *types.Var:
+		return obj.Origin()
+	case *types.Func:
+		return obj.Origin()
+	default:
+		return obj
+	}
+}
+
+// importLockGuardFacts imports lock guard facts for obj.
+//
+// For generic code, the `types.Object` identity observed during SSA analysis can
+// differ from the declaration object where facts were exported. When that
+// happens, we fall back to importing facts from the origin object.
+func (pc *passContext) importLockGuardFacts(obj types.Object, lgf *lockGuardFacts) {
+	pc.pass.ImportObjectFact(obj, lgf)
+	if len(lgf.GuardedBy) != 0 || lgf.AtomicDisposition != atomicDisallow {
+		return
+	}
+
+	// Fallback: import from the object's origin.
+	if orig := originObject(obj); orig != nil && orig != obj {
+		pc.pass.ImportObjectFact(orig, lgf)
+	}
+}
+
+// importLockFunctionFacts imports lock function facts for fn, falling back to
+// the origin object when fn is an instantiated generic function/method.
+func (pc *passContext) importLockFunctionFacts(fn *types.Func, lff *lockFunctionFacts) {
+	pc.pass.ImportObjectFact(fn, lff)
+	if lff.Ignore || len(lff.HeldOnEntry) != 0 || len(lff.HeldOnExit) != 0 || len(lff.ExcludedOnEntry) != 0 {
+		return
+	}
+	if orig := fn.Origin(); orig != nil && orig != fn {
+		pc.pass.ImportObjectFact(orig, lff)
+	}
+}

--- a/tools/checklocks/test/BUILD
+++ b/tools/checklocks/test/BUILD
@@ -19,6 +19,7 @@ go_library(
         "globals.go",
         "incompat.go",
         "inferred.go",
+        "generics.go",
         "locker.go",
         "methods.go",
         "parameters.go",

--- a/tools/checklocks/test/crosspkg/crosspkg.go
+++ b/tools/checklocks/test/crosspkg/crosspkg.go
@@ -24,3 +24,12 @@ var (
 	Foo   int
 	FooMu sync.Mutex
 )
+
+// GenericGuard is a generic type with a guarded field. This is used to verify
+// that facts exported by this package are correctly imported when another
+// package instantiates GenericGuard[T].
+type GenericGuard[T any] struct {
+	Mu sync.Mutex
+	// +checklocks:Mu
+	Value T
+}

--- a/tools/checklocks/test/generics.go
+++ b/tools/checklocks/test/generics.go
@@ -1,0 +1,91 @@
+// Copyright 2026 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"sync"
+
+	"gvisor.dev/gvisor/tools/checklocks/test/crosspkg"
+)
+
+var genericGlobalMu sync.Mutex
+
+type genericValueStruct[T any] struct {
+	mu sync.Mutex
+	// +checklocks:mu
+	value T
+}
+
+func (g *genericValueStruct[T]) set(v T) {
+	g.mu.Lock()
+	g.value = v
+	g.mu.Unlock()
+}
+
+func (g *genericValueStruct[T]) setUnlocked(v T) {
+	g.value = v // +checklocksfail
+}
+
+// +checklocks:g.mu
+func (g *genericValueStruct[T]) setLocked(v T) {
+	g.value = v
+}
+
+// genericMapStruct is a minimal reproduction of generic guarded fields.
+// See https://github.com/google/gvisor/issues/10372 and #11671.
+type genericMapStruct[K comparable] struct {
+	mu sync.Mutex
+	// +checklocks:mu
+	m map[K]struct{}
+}
+
+func (g *genericMapStruct[K]) add(k K) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	g.m[k] = struct{}{}
+}
+
+// genericUnlockedWrite should fail: writing without holding the guard.
+func genericUnlockedWrite[K comparable](g *genericMapStruct[K], k K) {
+	g.m[k] = struct{}{} // +checklocksfail
+}
+
+// +checklocks:genericGlobalMu
+func genericNeedsGlobalLock[T any](v T) {
+	_ = v
+}
+
+func genericCallsWithoutLock[T any](v T) {
+	genericNeedsGlobalLock(v) // +checklocksfail
+}
+
+func genericCallsWithLock[T any](v T) {
+	genericGlobalMu.Lock()
+	genericNeedsGlobalLock(v)
+	genericGlobalMu.Unlock()
+}
+
+// crossPkgGenericValid ensures that lock guard facts on generic types imported
+// from another package are still found when analyzing a generic use site.
+func crossPkgGenericValid[T any](g *crosspkg.GenericGuard[T], v T) {
+	g.Mu.Lock()
+	g.Value = v
+	g.Mu.Unlock()
+}
+
+// crossPkgGenericInvalidWrite should fail: writing without holding the guard.
+func crossPkgGenericInvalidWrite[T any](g *crosspkg.GenericGuard[T], v T) {
+	g.Value = v // +checklocksfail
+}

